### ISSUE TITLE
Donut Chart

### ISSUE THREAD

**Issue by truegold** [vdelachaux/SVG-with-Classes#6]
Hi Vincent,

Enjoying the extension of the SVG class. I love how easy it is to create a pie chart. But I have not been able to get the donut from it. I get very very strange results!

Anyway this would require both an inner and outer radius (I guess the differences between the radius is the thickness), a start angle and an end angle. Like this example I mentioned previously:

[https://observablehq.com/@haakenlid/svg-circle](url)

Could you show how to use the SVg class to create a donut chart?

Appreciate,
John...

**Comment by truegold** [comment]
Hi Vincent,

I noticed that you worked on implementing a donut. You've made more progress than I have using the svg class. Is this the "final" state? If so, it's a bit off. Here's what I see:

![dn1](https://github.com/vdelachaux/SVG-with-Classes/assets/25699211/e0a8dadd-380a-4e44-9a1d-77619965009d)

As you can see, the internal curves don't match to the inner circle.

Using classic SVG commands I have been able to get it working. Here's an example (still a bit messy) where the "thickness" (0-100) sets the width of the donut:

![dn2](https://github.com/vdelachaux/SVG-with-Classes/assets/25699211/143659d0-40ed-4d5e-80b4-8dd58c7fd43c)

The classic code looks like this in my class:

`
Function drawSegment($_o : Object)
	var $_p : Object
	var $PathID : Text
	var $posX; $posY : Real
	var $x2; $y2; $x3; $y3; $x4; $y4 : Real
	var $x2bis; $y2bis; $x4bis; $y4bis; $x5bis; $y5bis : Real
	
	$_p:=This._props
	$posX:=$_p.width/2
	$posY:=$_p.height/2
	
	$x2:=$posX+Int(Sin($_o.midAngle*Degree)*$_p.popoutRadius)
	$y2:=$posY-Int(Cos($_o.midAngle*Degree)*$_p.popoutRadius)
	
	// $_p.thickness
	$x2bis:=$posX+Int(Sin($_o.startAngle*Degree)*($_p.popoutRadius+$_p.innerRadius))
	$y2bis:=$posY-Int(Cos($_o.startAngle*Degree)*($_p.popoutRadius+$_p.innerRadius))
	
	$PathID:=SVG_New_path(This.svg.root; $x2bis; $y2bis; $_o.segColor; $_o.segColor; 0)  //0.5
	
	$x3:=$posX+Int(Sin($_o.startAngle*Degree)*($_p.outerRadius+$_p.popoutRadius))
	$y3:=$posY-Int(Cos($_o.startAngle*Degree)*($_p.outerRadius+$_p.popoutRadius))
	$x4:=$posX+Int(Sin($_o.endAngle*Degree)*($_p.outerRadius+$_p.popoutRadius))
	$y4:=$posY-Int(Cos($_o.endAngle*Degree)*($_p.outerRadius+$_p.popoutRadius))
	SVG_PATH_LINE_TO($PathID; $x3; $y3)
	
	If (($_o.endAngle-$_o.startAngle)>=180)
		$x4bis:=$posX+Int(Sin($_o.midAngle*Degree)*($_p.outerRadius+$_p.popoutRadius))
		$y4bis:=$posY-Int(Cos($_o.midAngle*Degree)*($_p.outerRadius+$_p.popoutRadius))
		SVG_PATH_ARC($PathID; $_p.outerRadius+$_p.popoutRadius; $_p.outerRadius+$_p.popoutRadius; $x4bis; $y4bis; 0; 0)
		SVG_PATH_ARC($PathID; $_p.outerRadius+$_p.popoutRadius; $_p.outerRadius+$_p.popoutRadius; $x4; $y4; 0; 0)
	Else 
		SVG_PATH_ARC($PathID; $_p.outerRadius+$_p.popoutRadius; $_p.outerRadius+$_p.popoutRadius; $x4; $y4; 0; 0)
	End if 
	
	$x5:=$posX+Int(Sin($_o.endAngle*Degree)*($_p.popoutRadius+$_p.innerRadius))
	$y5:=$posY-Int(Cos($_o.endAngle*Degree)*($_p.popoutRadius+$_p.innerRadius))
	SVG_PATH_LINE_TO($PathID; $x5; $y5)
	
	If (($_o.endAngle-$_o.startAngle)>=180)
		$x5bis:=$posX+Int(Sin($_o.midAngle*Degree)*($_p.popoutRadius+$_p.innerRadius))
		$y5bis:=$posY-Int(Cos($_o.midAngle*Degree)*($_p.popoutRadius+$_p.innerRadius))
		SVG_PATH_ARC($PathID; $_p.popoutRadius+$_p.innerRadius; $_p.popoutRadius+$_p.innerRadius; $x5bis; $y5bis; 0; 2)
		SVG_PATH_ARC($PathID; $_p.popoutRadius+$_p.innerRadius; $_p.popoutRadius+$_p.innerRadius; $x2bis; $y2bis; 0; 2)
	Else 
		SVG_PATH_ARC($PathID; $_p.popoutRadius+$_p.innerRadius; $_p.popoutRadius+$_p.innerRadius; $x2bis; $y2bis; 0; 2)
	End if 
	
	SVG_SET_ID($PathID; String($i))
	
	SVG_SET_OPACITY(This.svg.root; $_o.fillOpacity; $_o.strokeOpacity)
`

Not sure if this is helpful, but you understand the xml "magic" so much better than I do.

Appreciate,
John...

**Comment by truegold** [comment]
Hi Vincent,

Brilliant!

I have done some initial tests and it works.  Yea! I changed it to 12 slices (each 8.3333333333) and it looks pretty good. So it looks like the total normalizes to 100. The thickness seems to be in the 0.0001 -100 range. (Note: a thickness of 0 does not change and seems to then default to a minimum thickness)

`$chart.donut("1st"; 200; 200; 120; 0)`

![image](https://github.com/vdelachaux/SVG-with-Classes/assets/25699211/759313a3-64d0-4590-aa4c-b7e75545d4a6)

There is an issue with the margin. If I set the margin to 5:

`$chart.donut("1st"; 200; 200; 120; 0; 5)`

![image](https://github.com/vdelachaux/SVG-with-Classes/assets/25699211/b22365ca-653f-4555-a32d-6ac77e72af86)

Plese let me express my sincere appreciation for your work. This class is amazing!

John...
